walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,129 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Arrays;
 
+/**
+* The FindMeetingQuery class represents a meeting that we want to schedule given
+* the TimeRanges of events. 
+*/
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+  
+  /**
+  * Returns an unsorted ArrayList of TimeRanges that represent the times a group of attendees
+  * can meet.
+  */
+  private ArrayList<TimeRange> getAvailability(ArrayList<TimeRange> busyTimesForChosenAttendees) {
+    TimeRange currentBusyTime = busyTimesForChosenAttendees.get(0);
+    ArrayList<TimeRange> busyTimesForChosenAttendeesNoOverlaps = new ArrayList<>();
+    // Stores all TimeRanges in which attendees can meet.
+    ArrayList<TimeRange> freeTimesForChosenAttendees = new ArrayList<>();
+    
+    for (TimeRange time : busyTimesForChosenAttendees) {
+      // Checks whether our attendees have overlapping TimeRanges.
+      if (time.overlaps(currentBusyTime)) {
+        int start = currentBusyTime.start();
+        currentBusyTime = TimeRange.fromStartEnd(start, time.end() < currentBusyTime.end() ? currentBusyTime.end() : time.end(), false);
+      } else {
+        // Adds TimeRange when they don't conflict with each other.
+        busyTimesForChosenAttendeesNoOverlaps.add(currentBusyTime);
+        currentBusyTime = time;
+        }
+    }
+    // Adding last one, we now have the full calendar of not available time ranges. 
+    busyTimesForChosenAttendeesNoOverlaps.add(currentBusyTime);
+    TimeRange first = busyTimesForChosenAttendeesNoOverlaps.get(0);
+    TimeRange last = busyTimesForChosenAttendeesNoOverlaps.get(busyTimesForChosenAttendeesNoOverlaps.size() - 1);
+    
+    // Creates TimeRanges that include the start/end of the day.
+    TimeRange firstFreeTimeRange = TimeRange.fromStartEnd(TimeRange.START_OF_DAY, first.start(), false);
+    TimeRange lastFreeTimeRange = TimeRange.fromStartEnd(last.end(), TimeRange.END_OF_DAY, true);
+    
+    // Takes out TimeRanges that do not last enough time. 
+    if (firstFreeTimeRange.duration() >= request.duration()) {
+      freeTimeForChosenAttendees.add(firstFreeTimeRange);
+    }
+    if (lastFreeTimeRange.duration() >= request.duration()) {
+      freeTimesForChosenAttendees.add(lastFreeTimeRange);
+    }
+    // Checks whether we have enough TimeRanges to iterate through
+    if (busyTimesForChosenAttendeesNoOverlaps.size() > 1) {
```

Comment:
You could get rid of this "if" if you change the for loop below to iterate up to `busyTimesForChosenAttendeesNoOverlaps.size() - 2`. This would also fix this issue noted below about accessing a nonexistent index.